splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/CrossJoinOperation.java

```diff
@@ -0,0 +1,198 @@
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.derby.impl.sql.execute.operations;
+
+import com.splicemachine.EngineDriver;
+import com.splicemachine.access.api.SConfiguration;
+import com.splicemachine.client.SpliceClient;
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.services.loader.GeneratedMethod;
+import com.splicemachine.db.iapi.sql.Activation;
+import com.splicemachine.db.iapi.sql.execute.ExecRow;
+import com.splicemachine.db.impl.sql.compile.FromBaseTable;
+import com.splicemachine.derby.iapi.sql.execute.SpliceOperation;
+import com.splicemachine.derby.iapi.sql.execute.SpliceOperationContext;
+import com.splicemachine.derby.stream.function.*;
+import com.splicemachine.derby.stream.function.broadcast.BroadcastJoinFlatMapFunction;
+import com.splicemachine.derby.stream.function.broadcast.CogroupBroadcastJoinFunction;
+import com.splicemachine.derby.stream.function.broadcast.SubtractByKeyBroadcastJoinFunction;
+import com.splicemachine.derby.stream.iapi.DataSet;
+import com.splicemachine.derby.stream.iapi.DataSetProcessor;
+import com.splicemachine.derby.stream.iapi.OperationContext;
+import com.splicemachine.primitives.Bytes;
+import com.splicemachine.utils.SpliceLogUtils;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.List;
+
+
+public class CrossJoinOperation extends JoinOperation{
+    private static final long serialVersionUID=2l;
+    private static Logger LOG=Logger.getLogger(CrossJoinOperation.class);
+    protected int leftHashKeyItem;
+    protected int[] leftHashKeys;
+    protected int rightHashKeyItem;
+    protected int[] rightHashKeys;
+    protected long sequenceId;
+    protected static final String NAME = CrossJoinOperation.class.getSimpleName().replaceAll("Operation","");
+
+	@Override
+	public String getName() {
+			return NAME;
+	}
+
+    public CrossJoinOperation() {
+        super();
+    }
+
+    public CrossJoinOperation(SpliceOperation leftResultSet,
+                              int leftNumCols,
+                              SpliceOperation rightResultSet,
+                              int rightNumCols,
+                              int leftHashKeyItem,
```

Comment:
Do we need all these parameters? I don't think hashKeys make sense for a crossjoin where we are joining every row with each other.

Reply:
For join operations, restriction also contains non-equality conditions. So we must have hash keys for equality conditions. And these conditions can also pass to Spark's cross join implementation, which I think would be better to filter later.

Reply:
BTW, nested loop join don't have hash keys because equality conditions are pushed down to table scan (or other lower level operations).